### PR TITLE
CDAP-13088 MetadataReader and Writer context for programs

### PR DESCRIPTION
**Pull request by rohitsinha54**
- Part 2: APIs which will be exposed to the programs for reading/writing metadata. These APIs are similar to MetadataAdmin with minor improvements. We will refactor MetadataAdmin to use these interfaces itself.
- PR opened against metadata-rel50 which is the development branch for Metadata 5.0. We will periodically merge to develop.

Issue: https://issues.cask.co/browse/CDAP-13088
Build: https://builds.cask.co/browse/CDAP-DUT6306-1

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.api.metadata;
+
+import java.util.Map;
+import java.util.Set;
+import javax.ws.rs.NotFoundException;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: We don't use this class in API. Usually we define a hierarchy like `Exception` <- `MetadataException` <- `MetadataEntityNotFoundException`.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+   *
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Set<MetadataRecord> getMetadata(MetadataEntity metadataEntity) throws NotFoundException;
+
+  /**
+   * Returns a set of {@link MetadataRecord} representing all metadata (including properties and tags) for the specified
+   * {@link MetadataEntity} in the specified {@link MetadataScope}.
+   *
+   * @throws NotFoundException if the specified entity was not found
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: what if the given scope has no metadata, will it throw or return an empty set?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+   * @return all the properties for the specified {@link MetadataEntity} in both {@link MetadataScope#USER} and
+   * {@link MetadataScope#SYSTEM}
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Map<MetadataScope, Map<String, String>> getProperties(MetadataEntity metadataEntity) throws NotFoundException;
+
+  /**
+   * @return a {@link Map} representing the metadata of the specified {@link MetadataEntity} in the specified
+   * {@link MetadataScope}
+   * @throws NotFoundException if the specified entity was not found
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: same as above, if the given scope has no properties, will it throw or return an empty set?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  private final MetadataEntity metadataEntity;
+  private final MetadataScope scope;
+  private final Map<String, String> properties;
+  private final Set<String> tags;
+
+  /**
+   * Returns an empty {@link MetadataRecord} in the specified {@link MetadataScope}.
+   */
+  public MetadataRecord(MetadataEntity metadataEntity, MetadataScope scope) {
+    this(metadataEntity, scope, Collections.<String, String>emptyMap(), Collections.<String>emptySet());
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java]: You don't need the `<String, String>`. In Java 8, the compiler can infer it. Same for the `<String>`

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+   */
+  public MetadataRecord(MetadataRecord other) {
+    this(other.getMetadataEntity(), other.getScope(), other.getProperties(), other.getTags());
+  }
+
+  public MetadataRecord(MetadataEntity metadataEntity, MetadataScope scope, Map<String, String> properties,
+                        Set<String> tags) {
+    this.metadataEntity = metadataEntity;
+    this.scope = scope;
+    this.properties = properties;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java]: either do a defensive immutable copy or return an immutable one from the `getXXX` method.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+/**
+ * The context for emitting metadata from programs
+ */
+public interface MetadataWriterContext {
+
+  /**
+   * Adds the specified {@link Map} to the metadata of the specified {@link MetadataEntity metadataEntity}.
+   * Existing keys are updated with new values, newer keys are appended to the metadata.
+   *
+   * @throws NotFoundException if the specified entity was not found
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java]: can it be NotFound?? If an entity doesn't exists, shouldn't we just add it to metadata store??

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import javax.ws.rs.NotFoundException;
+
+/**
+ * The context for emitting metadata from programs
+ */
+public interface MetadataWriterContext {
+
+  /**
+   * Adds the specified {@link Map} to the metadata of the specified {@link MetadataEntity metadataEntity}.
+   * Existing keys are updated with new values, newer keys are appended to the metadata.
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java]: What do mean by newer keys are appended?? I though properties is just a `Map<String, String>`?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+   * @throws NotFoundException if the specified entity was not found
+   */
+  void addProperties(MetadataEntity metadataEntity, Map<String, String> properties) throws NotFoundException;
+
+  /**
+   * Adds the specified tags to specified {@link MetadataEntity}.
+   *
+   * @throws NotFoundException if the specified entity was not found
+   */
+  void addTags(MetadataEntity metadataEntity, String... tags) throws NotFoundException;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java]: Good to also have a method that takes `Iterable<String> tags`.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+  /**
+   * Removes all the user metadata (including properties and tags) for the specified {@link MetadataEntity}.
+   *
+   * @param metadataEntity the {@link MetadataEntity} to remove user metadata for
+   * @throws NotFoundException if the specified entity was not found
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java]: In practice, I don't think we'll be able to check at runtime, since these mutations will be emitted to TMS and executed asynchronously.

Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java]: True with new move it will not be possible. Will remove it. Also I dont think we can check for existence now regardless since metadata entity can be non cdap entities

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+ * the License.
+ */
+package co.cask.cdap.api.metadata;
+
+import java.util.Map;
+import java.util.Set;
+import javax.ws.rs.NotFoundException;
+
+/**
+ * The context for reading metadata from program
```
Comment (sagarkapare) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: From which phase of the program this functionality - `reading metadata` would be available? Will this be available in the `initialize` method or in the `prepareLaunch` which happens before provisioning or both?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java (hunk 1 above)
Comment (anew) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: At the same time it should extend NotFoundException? To allow generic HTTP error handling?

Comment (chtyim) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: We don't extend that one from javax.ws. For CDAP handlers, we check for class name to have `NotFoundException` in it.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java (hunk 2 above)
Comment (anew) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: It is not clear to me whether this should throw NotFoundExn. If I process a file in a Dataset and I am interested in its metadata, or I am interested in the metadata for a field of a dataset's schema, it may be that no metadata has been saved yet. But these are not registered CDAP entities... so the backend can't know whether this is an existing or non-existing entity. I think it should perhaps just return an empty set?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Map<MetadataScope, Set<String>> getTags(MetadataEntity metadataEntity) throws NotFoundException;
+
+  /**
+   * @return all the tags for the specified {@link MetadataEntity} in the specified {@link MetadataScope}
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Set<String> getTags(MetadataScope scope, MetadataEntity metadataEntity) throws NotFoundException;
+}
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: Is it necessary to have a separate method for tags and properties? Is there a performance impact? Or does it harm to just always return all metadata records?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+ */
+public interface MetadataReaderContext {
+
+  /**
+   * Returns a set of {@link MetadataRecord} representing all metadata (including properties and tags) for the specified
+   * {@link MetadataEntity} in both {@link MetadataScope#USER} and {@link MetadataScope#SYSTEM}.
+   *
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Set<MetadataRecord> getMetadata(MetadataEntity metadataEntity) throws NotFoundException;
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: why is this a set? MetadataRecord is already a collection of all tags and properties. So you will have only two entries (system and user)? Seems to make more sense to return a map from scope to record.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+  /**
+   * Returns a set of {@link MetadataRecord} representing all metadata (including properties and tags) for the specified
+   * {@link MetadataEntity} in the specified {@link MetadataScope}.
+   *
+   * @throws NotFoundException if the specified entity was not found
+   */
+  Set<MetadataRecord> getMetadata(MetadataScope scope, MetadataEntity metadataEntity) throws NotFoundException;
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: why is this a set? It will always be a singleton?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java (hunk 5 above)
Comment (anew) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java]: you must do a copy here, because you don't know whether the map and set passed in are mutable, and they might be changed later on.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataScope.java [hunk 14]
```diff
@@ -14,12 +14,12 @@
...
  */
 
-package co.cask.cdap.proto.metadata;
+package co.cask.cdap.api.metadata;
 
 import co.cask.cdap.api.annotation.Beta;
 
 /**
- * Represents the scope of metadata for an entity.
+ * Represents the scope of metadata for an metadata entity.
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataScope.java]: an/a

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java (hunk 6 above)
Comment (anew) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java]: agreed.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java (hunk 10 above)
Comment (anew) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: Didn't we say that we need a way to "announce" in the prepareLaunch() what meta data entities the program will read at runtime? And the runtime will then make all meta data for these entities available?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java (hunk 9 above)
Comment (anew) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java]: and that applies to the reader as well. See my comment there.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java (hunk 1 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: I don't think we should throw NotFoundException here. Since now we will support custom resource tagging we cannot check whether the resource exists or not. We can return an empty list in most cases if we don't find any metadata for the given key. How about this behavior?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+   */
+  void removeMetadata(MetadataEntity metadataEntity) throws NotFoundException;
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java]: will removal of this metadata have the effect that a subsequent read returns nothing? If writes are async, then we somehow need to cache the changes that have been applied in the context.

Comment (anew) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java]: and same for other changes

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java (hunk 10 above)
Comment (sagarkapare) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: I see. So this interface `MetadataReaderContext` is actually available to the program runtime context such as `initialize` method of the program, right? And only those metadata entities can be used for query which are actually announced in the `prepareLaunch`. Announcing the Metadata entity which will be needed during the runtime will be part of separate interface.

We should probably clearly mention that in the javadoc.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java (hunk 2 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: Yes, I agree 
https://github.com/caskdata/cdap/pull/9938#discussion_r171979065

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java (hunk 3 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: I will get rid of all NotFound the behavior will be to return empty set if the resource does not exists.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java (hunk 15 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java]: Yes. We will need to come up with how we will handle it during implementation. Depending on what we do I will update this doc accordingly.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java (hunk 11 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: We have separate methods in the our clients and rest apis so the ideas is taken from there for consistency of apis. Although, yes so many apis just to expose different part of the same data doesnt seem right. There is no performance impact as such because all these apis read the hbase table to get the metadata (including properties and tags) internally and then filter out whats needed as post processing. So we can just expose one which gives the MetadataRecord.

But now the idea is that we will announce the MetadataEntity whose metadata we want to be available in the context of initialize. If we do the above then we will only support making the whole MetadataRecord available in the context even if the user wants just the properties of tags. This might lead to filling up more memory than required.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java [hunk 16]
```diff
@@ -1,69 +0,0 @@
...
-package co.cask.cdap.api.metadata;
-
-import java.util.Map;
-import java.util.Set;
-import javax.ws.rs.NotFoundException;
-
-/**
- * The context for reading metadata from program
- */
-public interface MetadataReaderContext {
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: By removing this entirely, that means there no API to read meta data? I think we need a way to allow:
- in prepareLaunch, read and write metadata and mark it for use at runtime 
- at runtime (including initialize/destroy), read metadata marked in prepareLaunch, submit writes for meta data
- in postExecution, read and write metadata (and that may include meta data changed at runtime)

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRegistrant.java [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+ * the License.
+ */
+package co.cask.cdap.api.metadata;
+
+import co.cask.cdap.api.ProgramLifecycle;
+import co.cask.cdap.api.RuntimeContext;
+
+/**
+ * Class which registers metadata for programs so
+ * that they are available in {@link co.cask.cdap.api.ProgramLifecycle#initialize(RuntimeContext)} and
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRegistrant.java]: also at runtime?

Comment (anew) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRegistrant.java]: which API will programs use to read meta data that was made available this way?

File: cdap-proto/src/test/java/co/cask/cdap/proto/audit/AuditMessageTest.java [hunk 18]
```diff
@@ -49,7 +49,8 @@
     .registerTypeAdapter(EntityId.class, new EntityIdTypeAdapter())
     .create();
 
-  private static final Type MAP_TYPE = new TypeToken<Map<String, Object>>() { }.getType();
+  private static final Type MAP_TYPE = new TypeToken<Map<String, Object>>() {
```
Comment (anew) [inline cdap-proto/src/test/java/co/cask/cdap/proto/audit/AuditMessageTest.java]: unnecessary change?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+  public MetadataRecord(MetadataRecord other) {
+    this(other.getMetadataEntity(), other.getScope(), other.getProperties(), other.getTags());
+  }
+
+  public MetadataRecord(MetadataEntity metadataEntity, MetadataScope scope, Map<String, String> properties,
+                        Set<String> tags) {
+    this.metadataEntity = metadataEntity;
+    this.scope = scope;
+    this.properties = Collections.unmodifiableMap(new HashMap<>(properties));
+    this.tags = tags;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java]: I believe you do need to copy the tags.

Comment (chtyim) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java]: Also, is the contract for tags be an unordered set of string?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRegistrant.java [hunk 20]
```diff
@@ -0,0 +1,41 @@
...
+ * {@link ProgramLifecycle#destroy()}
+ */
+public interface MetadataRegistrant {
+
+  /**
+   * Register the {@link MetadataRecord} (including properties and tags) for the specified {@link MetadataEntity}
+   * in both {@link MetadataScope#USER} and {@link MetadataScope#SYSTEM} if the metadata entity exists and has any
+   * metadata associated with it.
+   */
+  void registerMetadata(MetadataEntity metadataEntity);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRegistrant.java]: So register method won't return the metadata?? Then what interface is responsible for actually reading data (either from the metadata store directly at prepare launch time, or from the captured metadata at runtime).

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+public interface MetadataWriterContext {
+
+  /**
+   * Adds the specified {@link Map} to the metadata of the specified {@link MetadataEntity metadataEntity}.
+   * Existing keys will be updated with new values.
+   */
+  void addProperties(MetadataEntity metadataEntity, Map<String, String> properties);
+
+  /**
+   * Adds the specified tags to specified {@link MetadataEntity}.
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java]: In your above code, tags are represented as `Set<String>`, this means duplicated tags will be removed, right? If that's the case, I think it's better for this javadoc to specify that behavior.

Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataWriterContext.java]: Since

**Comment by rohitsinha54** [comment]
@chtyim @anew : Please have a look. Had a good discussion with Terence to get clarity on the `prepareLaunch` and `initialize` difference and what APIs will look like in both cases. Lastly, the idea is that we don't need any specific announce API as such. The same reader api will be available in both cases. During the `prepareLaunch` the reader can read all the metadata and during`initialize` only which was made available during the runtime.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+
+  public MetadataRecord(MetadataEntity metadataEntity, MetadataScope scope, Map<String, String> properties,
+                        Set<String> tags) {
+    this.metadataEntity = metadataEntity;
+    this.scope = scope;
+    this.properties = Collections.unmodifiableMap(new HashMap<>(properties));
+    this.tags = Collections.unmodifiableSet(new HashSet<>(tags));
+  }
+
+  public MetadataEntity getMetadataEntity() {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java]: You need javadoc for public methods in cdap-api.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * The context for reading metadata from program
+ */
+public interface MetadataReaderContext {
+
+  /**
+   * Returns a set of {@link MetadataRecord} representing all metadata (including properties and tags) for the specified
+   * {@link MetadataEntity} in both {@link MetadataScope#USER} and {@link MetadataScope#SYSTEM}.
+   */
+  Set<MetadataRecord> getMetadata(MetadataEntity metadataEntity);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: Shouldn't the result be `Map<MetadataScope, MetadataRecord>`?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+   * Returns a set of {@link MetadataRecord} representing all metadata (including properties and tags) for the specified
+   * {@link MetadataEntity} in the specified {@link MetadataScope}.
+   */
+  Set<MetadataRecord> getMetadata(MetadataScope scope, MetadataEntity metadataEntity);
+
+  /**
+   * @return all the properties for the specified {@link MetadataEntity} in both {@link MetadataScope#USER} and
+   * {@link MetadataScope#SYSTEM}
+   */
+  Map<MetadataScope, Map<String, String>> getProperties(MetadataEntity metadataEntity);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: So the getProperties and getTags are just convenient methods? I feel like it's better not to have them in the API, but just let the user use the `MetadataRecord`.

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+   * Returns a set of {@link MetadataRecord} representing all metadata (including properties and tags) for the specified
+   * {@link MetadataEntity} in both {@link MetadataScope#USER} and {@link MetadataScope#SYSTEM}.
+   */
+  Set<MetadataRecord> getMetadata(MetadataEntity metadataEntity);
+
+  /**
+   * Returns a set of {@link MetadataRecord} representing all metadata (including properties and tags) for the specified
+   * {@link MetadataEntity} in the specified {@link MetadataScope}.
+   */
+  Set<MetadataRecord> getMetadata(MetadataScope scope, MetadataEntity metadataEntity);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataReaderContext.java]: Why the result is a `Set`? Isn't that there should be just one `MetadataRecord` for a given entity in a given scope?

File: cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Represents the complete metadata of a {@link MetadataEntity} including its properties, tags for a given
+ * {@link MetadataScope}
+ */
+public class MetadataRecord {
+  private final MetadataEntity metadataEntity;
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/metadata/MetadataRecord.java]: Should the entity be part of the record? If it does, that means you can't compare the metadata of two different entities directly.

**Review by chtyim — COMMENTED** [review]
Just one question left.

**Comment by chtyim** [comment]
LGTM
